### PR TITLE
[spark] Enhance spark push down filter

### PR DESCRIPTION
**Pull request by Zouxxyy**
### Purpose

Linked issue: close #xxx

### CODE REVIEW

File: paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/PaimonSparkTestBase.scala [hunk 1]
```diff
@@ -90,4 +90,8 @@ class PaimonSparkTestBase extends QueryTest with SharedSparkSession with WithTab
   def loadTable(tableName: String): AbstractFileStoreTable = {
     catalog.getTable(Identifier.create(dbName0, tableName)).asInstanceOf[AbstractFileStoreTable]
   }
+
+  def explain(sql: String): String = {
+    spark.sql(s"EXPLAIN $sql").collect().map(_.mkString).mkString
```
Comment (YannByron) [inline paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/PaimonSparkTestBase.scala:95]: please use `queryExecution` to determine logical/physical plan.

File: paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/SparkPushDownTest.scala [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+  test(s"Paimon push down: apply partition filter push down with non-partitioned table") {
+    spark.sql(s"""
+                 |CREATE TABLE T (id INT, name STRING, pt STRING)
+                 |TBLPROPERTIES ('primary-key'='id, pt', 'bucket'='2')
+                 |""".stripMargin)
+
+    spark.sql("INSERT INTO T VALUES (1, 'a', 'p1'), (2, 'b', 'p1'), (3, 'c', 'p2')")
+
+    assertThat(spark.sql("SELECT * FROM T WHERE pt = 'p1'").queryExecution.optimizedPlan.exists {
```
Comment (YannByron) [inline paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/SparkPushDownTest.scala:35]: Assertions.assertTrue.

File: paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/SparkPushDownTest.scala [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    spark.sql("INSERT INTO T VALUES (1, 'a', 'p1'), (2, 'b', 'p1'), (3, 'c', 'p2'), (4, 'd', 'p3')")
+
+    // partition filter push down did not hit
+    assertThat(spark.sql("SELECT * FROM T WHERE id = '1'").queryExecution.optimizedPlan.exists {
+      case Filter(_: Expression, _) => true
+      case _ => false
+    }).isTrue
+    checkAnswer(spark.sql("SELECT * FROM T WHERE id = '1'"), Row(1, "a", "p1") :: Nil)
+
+    assertThat(
```
Comment (YannByron) [inline paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/SparkPushDownTest.scala:62]: please add some comments for this test.

File: paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/SparkPushDownTest.scala [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+    assertThat(spark.sql("SELECT * FROM T WHERE pt < 'p3'").queryExecution.optimizedPlan.exists {
+      case Filter(_: Expression, _) => true
+      case _ => false
+    }).isFalse
+    checkAnswer(
+      spark.sql("SELECT * FROM T WHERE pt < 'p3'"),
+      Row(1, "a", "p1") :: Row(2, "b", "p1") :: Row(3, "c", "p2") :: Nil)
+  }
+
```
Comment (YannByron) [inline paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/SparkPushDownTest.scala:104]: maybe define two private methods (`checkFilterExists` and `checkFilterEquals`). These can be reused.

File: paimon-common/src/main/java/org/apache/paimon/predicate/PartitionPredicateVisitor.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    @Override
+    public Boolean visit(LeafPredicate predicate) {
+        return partitionKeys.contains(predicate.fieldName());
+    }
+
+    @Override
+    public Boolean visit(CompoundPredicate predicate) {
+        for (Predicate child : predicate.children()) {
```
Comment (YannByron) [inline paimon-common/src/main/java/org/apache/paimon/predicate/PartitionPredicateVisitor.java:39]: Thinking about a case that one child of `CompoundPredicate` is also `CompoundPredicate`.

**Comment by YannByron** [comment]
+1

**Review by JingsongLi — APPROVED** [review]
+1

**Comment by YannByron** [comment]
https://github.com/apache/incubator-paimon/issues/2404

**Pull request merged**